Documentation update: other (#732)
flame-engine/flame: doc/palette.md

# Palette

Throughout your game you are going to need to use colors in lots of places. There are two classes on
`dart:ui` that can be used, `Color` and `Paint`.

The `Color` class is nothing but a wrapper over a single simple ARGB color in hexadecimal integer
format. So to create a `Color` object you just need to pass in the color as an integer in the ARGB
format.

You can use Dart's hexadecimal notation to make it really easy; for instance: `0xFF00FF00` is fully
opaque green (the 'mask' would be `0xAARRGGBB`). Do note the first two hexadecimal digits are for
the alpha channel (transparency), unlike on regular (non-A) RGB. The max(FF = 255) for the two first
digits means fully opaque, and the min (00 = 0) means fully transparent.

In the material flutter package there is a color enum to make it easy to use common colors:

```dart
import 'package:flutter/material.dart';

Color black = Colors.black;
```

Some more complex methods might also take a `Paint` object, which is a more complete option that
allows you to define more options related to stroke, colors, filters and blends.

> However, normally when using even the more complex APIs, you just want an instance of a `Paint`

Reviewer: Maybe add a `**Note**:` that we recommend that you dont constantly create new `Paint` objects each render frame?

Author: Done

the alpha channel (transparency), unlike on regular (non-A) RGB. The max(FF = 255) for the two first
digits means fully opaque, and the min (00 = 0) means fully transparent.

In the material flutter package there is a `Colors` class that provides common colors as constants:

```dart
import 'package:flutter/material.dart';

const black = Colors.black;
```

Some more complex methods might also take a `Paint` object, which is a more complete option that
allows you to define more options related to stroke, colors, filters and blends.
However, normally when using even the more complex APIs, you just want an instance of a `Paint`
object representing just a single simple plain solid color.

You can create such an object like this:

```dart
Paint green = Paint()..color = const Color(0xFF00FF00);
```

To help you with this and to also keep your game's color palette consistent, Flame adds the `Palette`
class. You can use it to easily access both `Color`s and `Paint`s where needed and also to define
the colors your game use as constants, so you don't get those mixed up.

The `BasicPalette` class is an example of what a palette can look like, and adds black and white as

> colors. So to use black or white you can access those directly from the BasicPalette; for example,

Reviewer: ```suggestion
colors. So to use black or white you can access those directly from the `BasicPalette`; for example,
```

# Palette

Throughout your game you are going to need to use colors in lots of places. There are two classes on
`dart:ui` that can be used, `Color` and `Paint`.

The `Color` class represents a ARGB color in a hexadecimal integer

> format. So to create a `Color` instance you just need to pass the color as an integer in the ARGB

Reviewer 2: ```suggestion
format. So to create a `Color` instance, you just need to pass the color as an integer in the ARGB
```

format.

You can use Dart's hexadecimal notation to make it really easy; for instance: `0xFF00FF00` is fully

> opaque green (the 'mask' would be `0xAARRGGBB`). 

Reviewer 2: ```suggestion
opaque green (the "mask" would be `0xAARRGGBB`). 
```

**Note**: The first two hexadecimal digits are for
the alpha channel (transparency), unlike on regular (non-A) RGB. The max(FF = 255) for the two first
digits means fully opaque, and the min (00 = 0) means fully transparent.

> In the material flutter package there is a `Colors` class that provides common colors as constants:

Reviewer 2: ```suggestion
In the Material Flutter package there is a `Colors` class that provides common colors as constants:
```

```dart

> import 'package:flutter/material.dart';

Reviewer 2: show Colors

just a good practice to avoid clashes

const black = Colors.black;
```

> Some more complex methods might also take a `Paint` object, which is a more complete option that
> allows you to define more options related to stroke, colors, filters and blends.

Reviewer 2: ```suggestion
Some more complex methods might also take a `Paint` object, which is a more complete structure that
allows you to configure aspects related to stroke, colors, filters and blends.
```

However, normally when using even the more complex APIs, you just want an instance of a `Paint`
object representing just a single simple plain solid color.

**Note:** we don't recommend that you create a new `Paint` object every time you need a specific
`Paint`, since it could potentially lead to a lot of unnecessary objects being created. A better way
is to either define the `Paint` object somewhere and re-use it, or to use the `Palette` class to

> define all the colors that you want to use in your game.

Reviewer 2: However, also note that the Paint class is mutable, unlike Color.

}
```

> A `PaletteEntry` is a const class that holds information of a color and it has the following

Reviewer: ```suggestion
A `PaletteEntry` is a constant class that holds information of a color and it has the following
```

members:

 - `color`: returns the `Color` specified

>  - `paint`: creates a new `Paint` with the color specified. `Paint` is a non-const class, so this

Reviewer: ```suggestion
 - `paint`: creates a new `Paint` with the color specified. `Paint` is a non-constant class, so this
```

Author: I don't think it makes it clearer to spell it out, since the keyword in dart is `const`, but maybe we should backtick it?

A `PaletteEntry` is a const class that can hold information of a color, and has two attributes:

 - `color`: returns the `Color` specified

>  - `paint`: creates a new `Paint` with the color specified. `Paint` is a non-const class, so this
>   attribute actually creates a brand new instance every time it's called. It's safe to cascade
>   mutations to this.

Reviewer 3: ```suggestion
 - `paint`: creates a new `Paint` with the color specified. `Paint` is a non-const class, so this
  method actually creates a brand new instance every time it's called. It's safe to cascade
  mutations to this.
```
